src/utp_stream.cpp

```diff
@@ -242,7 +242,14 @@ struct packet
 // its size is "m_read_buffer_size". The buffer we spill over
 // into when the user provided buffer is full or when there
 // is none, is "m_receive_buffer" and "m_receive_buffer_size"
-// respectively.
+// respectively. 
+
+// "m_receive_buffer_capacity" marks the maximum amount of
+// bytes receive buffer should store at once. If we
+// want to reduce this limit during runtime, we must be careful
+// not to reduce it below the window size we advertised
+// to our peer earlier. The number of bytes we want to 
+// reduce receive buffer by is stored in "m_receive_buffer_surplus". 
```

Comment:
every time we send an ACK, we advertise the whole window. In order to decrease it, we must always wait for the upper layer to consume bytes (which normally would free up space in the receive buffer). Instead of freeing it up, we shrink the receive buffer by the consumed amount (or less if we're not shrinking it that much).

As I see it, there are two ways of implementing this. (1) is to have two separate window sizes, one that we're aiming for and one that we've advertised (the effective window), and then at the appropriate time inch the effective window closer to the target.

However, I think the (2) is a bit simpler (and less intrusive). We keep the current effective window size, but in addition we have a *delta*, indicating how much we want to lower it by. Every time the upper layer consumers bytes, we reduce the window size by min(consumed, want_to_reduce_wnd) and also reduce the ``want_to_reduce_wnd`` by the same amount.

When we want to increase the size, we first "eat into" the ``want_to_reduce`` and then increase the effective window size.

The main reason I think that's simpler is that the existing window size variable does not change meaning, and any code interacting with it can stay the same. My first reading of this code was that you've implemented (2) and ``m_receive_buffer_surplus`` is this counter I called ``want_to_reduce_wnd`` above.